accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1326,7 +1326,7 @@ fresh, non-late, final variable `v` is created. An initializing formal
 argument passed to this formal. An initializer list element of the
 form `id = e` or `this.id = e` is evaluated by evaluating `e` to an
 object `o` and binding `v` to `o`.  During the execution of the
-constructor body, `this` and `id` are bound to the value of `v`.  The
+constructor body, `this` is bound to the value of `v`.  The
```

Comment:
Why would it be an improvement to eliminate the rule that the representation variable is bound to the representation object? As far as I can see this means that a conforming implementation could bind the representation variable to an arbitrary object (OK, an arbitrary object whose run-time type is a subtype of the actual value of the representation type).

Reply:
It says `id` is bound to the value `v`. And `id` is not the representation variable, that's the *name* of the representation variable. 
Binding the representation variable to `v` is fine, if that's what it wants to say.
Binding an identifier to a value during execution of code means something else.

Since theree can be other things in scope with the same name, and I read this as saying that the *identifier* is bound to the value `v` while executing the constructor body. (I really couldn't read it any other way.)
Which it isn't.

Reply:
> It says `id` is bound to the value `v`

No, it says that `id` is bound to the value _of_ `v`. As is common in the language specification we introduce a fresh local variable (here: `v`) whose allocation is unspecified, but it occurs in the context of the semantics of the evaluation of an expression `e`, so it's presumably always possible to allocate it as an additional local variable in the scope which is the current scope for `e`.

This is slightly magic in cases like `var x = e;` because there is no syntactic support for allocating local variables in the initializing expression of a variable. On the other hand, the language specification does use let expressions in order to specify the semantics of certain constructs.

So let's say that we can in fact create a fresh variable like that. (Otherwise it's certainly a broader discussion if we insist that the language specification should be rewritten to avoid using this kind of local variables to specify the semantics of some expressions, and we should handle that separately.)

It is also common for our specification documents (including the language specification) to refer to a variable (of any kind) using its name. In some cases we say "the variable 'someName' is bound to ...", but in other cases we just say "'someName' is bound to ...".

We do not have the concept that an identifier is bound to anything.

I would not have a problem with "`this` and the variable `id` are bound to ...", but it does seem somewhat verbose, considering that we don't always do so.

> Binding the representation variable to `v` is fine

We never bind a variable to a variable. In this case I really don't think we have a habit of abbreviating "the value of `v`" to "`v`", and I wouldn't want to start doing that. (Well, why would that be worse? I guess it's because there is no way we could misunderstand "bind `v` to ..." to mean that we're binding the identifier to an object, but both the variable and the value of the variable are run-time entities, so that's more likely to be an actual source of confusion).

> Binding an identifier to a value during execution of code means something else.

We do have the notion of a run-time namespace (as well as a compile-time namespace). A run-time namespace could actually (if we squint) be considered to bind an identifier to a storage location, that is, it binds the identifier to a variable (and that storage location would in turn hold a pointer to an object, except that we don't specify the run-time semantics with that much detail, and it doesn't have to be a pointer, e.g., it could be a SmallInteger).

Nevertheless, the language specification and feature specifications don't usually include these details. I'd prefer to say that we (slightly magically) obtain a fresh variable, and that variable is bound to an object.

That's how we talk about objects that don't occur as the value of any variable when we specify the semantics of a construct.

> Since theree can be other things in scope with the same name ...

That's a good point! We need to mention the scopes where the representation name resolves to the representation variable, because there could be other variables (parameters or locals) shadowing it. The specification of the semantics of an initializing formal is still fine, and so is the specification of the semantics of initializer list elements of the form `id = e` and `this.id = e` (they will both initialize the representation variable, even in the case where there is a formal parameter named `id`).

But the part about the binding of the representation variable should not refer to its name.

So maybe:

> During the execution of the constructor body, `this` is bound to the value of `v`, and so is the representation variable.